app/src/main/java/com/atomicrobot/carbon/ui/scanner/ScannerViewModel.kt

```diff
@@ -0,0 +1,209 @@
+package com.atomicrobot.carbon.ui.scanner
+
+import android.Manifest
+import android.app.Application
+import android.os.SystemClock
+import androidx.camera.core.CameraSelector
+import androidx.camera.core.ImageAnalysis
+import androidx.camera.core.ImageProxy
+import androidx.camera.lifecycle.ProcessCameraProvider
+import androidx.core.content.ContextCompat
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.atomicrobot.carbon.util.Common
+import com.google.android.gms.tasks.OnFailureListener
+import com.google.android.gms.tasks.OnSuccessListener
+import com.google.android.gms.tasks.TaskExecutors
+import com.google.mlkit.vision.barcode.BarcodeScanner
+import com.google.mlkit.vision.barcode.BarcodeScannerOptions
+import com.google.mlkit.vision.barcode.BarcodeScanning
+import com.google.mlkit.vision.barcode.common.Barcode
+import com.google.mlkit.vision.common.InputImage
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.launch
+import java.util.concurrent.ExecutionException
+
+class ScannerViewModel(private val app: Application) :
+    ViewModel(),
+    ImageAnalysis.Analyzer,
+    OnSuccessListener<List<Barcode>>,
+    OnFailureListener {
+
+    private var options: BarcodeScannerOptions = BarcodeScannerOptions.Builder()
+        .setBarcodeFormats(
+            Barcode.FORMAT_EAN_13,
+            Barcode.FORMAT_UPC_E,
+            Barcode.FORMAT_CODE_128,
+            Barcode.FORMAT_QR_CODE
+        )
+        .build()
+
+    private val barcodeScanner: BarcodeScanner = BarcodeScanning.getClient(options)
+
+    private val hasCameraPermission: Boolean
+        get() = Common.hasPermission(app, Manifest.permission.CAMERA)
+
+    private val _cameraPermissionState: MutableStateFlow<Boolean> =
+        MutableStateFlow(hasCameraPermission)
+
+    private val _cameraSelectorState: MutableStateFlow<CameraSelector> =
+        MutableStateFlow(
+            CameraSelector
+                .Builder()
+                .requireLensFacing(CameraSelector.LENS_FACING_BACK)
+                .build()
+        )
+
+    private val _barcodeStateState = MutableStateFlow(BarcodeAnalysisState())
+
+    val cameraPermissionState = _cameraPermissionState.asStateFlow()
+
+    val cameraSelectorState = _cameraSelectorState.asStateFlow()
+
+    val barcodeOverlayState = _barcodeStateState.asStateFlow()
+
+    private val currentBarcode: Barcode?
+        get() = _barcodeStateState.value.barcode
+
+    private val isFrontFacing
+        get() = _cameraSelectorState.value.lensFacing == CameraSelector.LENS_FACING_FRONT
+
+    private val isBackFacing
+        get() = isFrontFacing
+
+    private val isImageFlipped
+        get() = isFrontFacing
+
```

Comment:
I think this is okay for our app. But maybe we don't need this variable? We can just check if it's front facing and we know that it is flipped in our app?
(thinking that if the user was able to flip the image for each camera then the check would need to be different. I know we don't do that for this app,  but this logic/naming doesn't seem to be helpful? Or maybe we just leave a comment next to it explaining)

Reply:
Removed